File: day1/rooks/solutions/nm2.cpp
// O((NM)^2))

#include <vector>
#include <queue>
#include "rooks.h"

std::vector<std::vector<int>> calculate_moves(std::vector<std::vector<int>> A, int K) {
    int N = A.size();
    int M = A[0].size();

    std::vector<std::vector<int>> dist(N, std::vector<int>(M, -1));

    std::queue<std::pair<int, int>> q;
    int start_r, start_c;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (A[i][j] == 1) {
                start_r = i;
                start_c = j;
                break;
            }
        }
    }

    dist[start_r][start_c] = 0;
    q.push({start_r, start_c});

    while (!q.empty()) {
        std::pair<int, int> current = q.front();
        q.pop();
        int r = current.first;
        int c = current.second;

        for (int r_prime = 0; r_prime < N; ++r_prime) {
            for (int c_prime = 0; c_prime < M; ++c_prime) {
                if ((r == r_prime || c == c_prime) && (r != r_prime || c != c_prime)) {
                    if (A[r_prime][c_prime] - A[r][c] > 0 && A[r_prime][c_prime] - A[r][c] <= K) {
                        if (dist[r_prime][c_prime] == -1) {
                            dist[r_prime][c_prime] = dist[r][c] + 1;
                            q.push({r_prime, c_prime});
                        }
                    }
                }
            }
        }
    }

    return dist;
}

File: day1/rooks/solutions/nm2.java
import java.util.LinkedList;
import java.util.Queue;

public class rooks {

    public static int[][] calculate_moves(int[][] A, int K) {
        int N = A.length;
        int M = A[0].length;

        int[][] dist = new int[N][M];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                dist[i][j] = -1;
            }
        }

        Queue<int[]> q = new LinkedList<>();
        int startR = -1, startC = -1;

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (A[i][j] == 1) {
                    startR = i;
                    startC = j;
                    break;
                }
            }
            if (startR != -1) {
                break;
            }
        }

        dist[startR][startC] = 0;
        q.add(new int[]{startR, startC});

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int r = current[0];
            int c = current[1];

            for (int rPrime = 0; rPrime < N; ++rPrime) {
                for (int cPrime = 0; cPrime < M; ++cPrime) {
                    if ((r == rPrime || c == cPrime) && (r != rPrime || c != cPrime)) {
                        if (A[rPrime][cPrime] - A[r][c] > 0 && A[rPrime][cPrime] - A[r][c] <= K) {
                            if (dist[rPrime][cPrime] == -1) {
                                dist[rPrime][cPrime] = dist[r][c] + 1;
                                q.add(new int[]{rPrime, cPrime});
                            }
                        }
                    }
                }
            }
        }
        return dist;
    }
}